include/ic_gameobject.h

```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include "ic_components.h"
+
+#include <memory>
+#include <vector>
+
+namespace IC {
+    class GameObject {
+    public:
+        GameObject(std::string name);
+        ~GameObject();
+
+        std::vector<std::shared_ptr<GameObjectComponent>> &Components() { return _components; }
+        std::shared_ptr<Transform> &GetTransform() { return _transform; }
+        template <typename T> void AddComponent(const T &component) {
+            _components.push_back(std::make_shared<T>(component));
+        }
+        template <typename T> GameObjectComponent *GetComponent() {
+            for (auto component : _components) {
+                if (typeid(*component) == typeid(T)) {
+                    return component.get();
+                }
+            }
+        }
+
+        void Gui();
+
+    private:
+        std::string _name;
+        std::shared_ptr<Transform> _transform;
+        std::vector<std::shared_ptr<GameObjectComponent>> _components;
```

Comment:
Idea: If we're going with the `typeid` approach for identifying components and do not desire the ability to have duplicates, we could instead use a map or unordered_map to store components with `std::type_index` as the key: https://en.cppreference.com/w/cpp/types/type_index

Reply:
^^This is the approach I took in the [2D game engine](https://github.com/Colbydude/2DGameEngine) a while back (see [Entity.h](https://github.com/Colbydude/2DGameEngine/blob/main/src/Entity.h)). :)